src/responses/Albany_QuadraticLinearOperatorBasedResponseFunction.cpp

```diff
@@ -0,0 +1,343 @@
+#include "MatrixMarket_Tpetra.hpp"
+
+Albany::QuadraticLinearOperatorBasedResponseFunction::
+QuadraticLinearOperatorBasedResponseFunction(const Teuchos::RCP<const Albany::Application> &app,
+    Teuchos::ParameterList &responseParams) :
+  SamplingBasedScalarResponseFunction(app->getComm()),
+  app_(app)
+{
+  coeff_ = responseParams.get<double>("Scaling Coefficient");
+  field_name_ = responseParams.get<std::string>("Field Name");
+  file_name_A_ = responseParams.get<std::string>("Linear Operator File Name");
+  file_name_D_ = responseParams.get<std::string>("Diagonal Scaling File Name");
+}
+
+Albany::QuadraticLinearOperatorBasedResponseFunction::
+~QuadraticLinearOperatorBasedResponseFunction()
+{
+}
+
+unsigned int
+Albany::QuadraticLinearOperatorBasedResponseFunction::
+numResponses() const 
+{
+  return 1;
+}
+
+void
+Albany::QuadraticLinearOperatorBasedResponseFunction::
+loadLinearOperator() {
+  Teuchos::RCP<const Thyra_Vector> field = app_->getDistributedParameterLibrary()->get(field_name_)->vector();
+  const Tpetra_GO globalNumElts = field->space()->dim();
+  Teuchos::RCP<const Tpetra_Map> rowMap = Albany::getTpetraMap(field->space());
+
+  Teuchos::RCP<const Tpetra_Map> colMap;
+  Teuchos::RCP<const Tpetra_Map> domainMap = rowMap;
+  Teuchos::RCP<const Tpetra_Map> rangeMap = rowMap;
+  typedef Tpetra::MatrixMarket::Reader<Tpetra_CrsMatrix> reader_type;
+
+  bool mapIsContiguous =
+      (rowMap->getMaxAllGlobalIndex()+1-rowMap->getMinAllGlobalIndex() ==
+      rowMap->getGlobalNumElements());
+
+  TEUCHOS_TEST_FOR_EXCEPTION (!mapIsContiguous, std::runtime_error,
+                              "Error! Row Map needs to be contigous for the Matrix reader to work.\n");
+
+  auto tpetra_mat =
+      reader_type::readSparseFile (file_name_A_, rowMap, colMap, domainMap, rangeMap);
+
+  auto tpetra_diag_mat =
+      reader_type::readSparseFile (file_name_D_, rowMap, colMap, domainMap, rangeMap);
+  Teuchos::RCP<Tpetra_Vector> tpetra_diag_vec = Teuchos::rcp(new Tpetra_Vector(rowMap));
+  tpetra_diag_mat->getLocalDiagCopy (*tpetra_diag_vec);
+
+  A_ = Albany::createThyraLinearOp(tpetra_mat);
+  D_ = Albany::createThyraVector(tpetra_diag_vec);
+  vec1_ = Thyra::createMember(A_->range());
+  vec2_ = Thyra::createMember(A_->range());
+}
+
+void
+Albany::QuadraticLinearOperatorBasedResponseFunction::
+evaluateResponse(const double /*current_time*/,
+    const Teuchos::RCP<const Thyra_Vector>& /*x*/,
+    const Teuchos::RCP<const Thyra_Vector>& /*xdot*/,
+    const Teuchos::RCP<const Thyra_Vector>& /*xdotdot*/,
+		const Teuchos::Array<ParamVec>& /*p*/,
+		const Teuchos::RCP<Thyra_Vector>& g)
+{
+
+  if(A_.is_null())
+    loadLinearOperator();
+
+  Teuchos::RCP<const Thyra_Vector> field = app_->getDistributedParameterLibrary()->get(field_name_)->vector();
+
+  // A p
+  A_->apply(Thyra::EOpTransp::NOTRANS, *field, vec1_.ptr(), 1.0, 0.0);
+
+  // coeff inv(D) A p
+  vec2_->assign(0.0);
+  Thyra::ele_wise_divide( coeff_, *vec1_, *D_, vec2_.ptr() );
```

Comment:
Thyra::VectorBase has a method to compute a weighted 2 norm. You could just do (here and in the other evaluate methods)
```
g->assign (std::pow(vec1->norm_2(*D_),2));
```
and do everything in one loop.

Reply:
NVM, the Tpetra adapter impl does precisely what you are doing.

Reply:
here I need to divide by `D`, not multiply by `D`.

Reply:
Ah, right.